Add C API guides (#2796)
Qiskit/documentation: docs/guides/c-extension-for-python.mdx

---
title: Extend Qiskit in Python with C
description: How to build a C extension for your Qiskit workflow in Python
---

# Extend Qiskit in Python with C

To accelerate your Qiskit Python programs with C, you can use the Qiskit C extension for Python.
This requires additional steps to the standalone C usage; for more details, see the [Install the Qiskit C API guide](/guides/install-c-api).

<Admonition type="note">
  These instructions have only been tested on UNIX-based systems. Windows instructions are in progress._
</Admonition>

> 

Reviewer: I think we need to include huge warnings here that building a C extension against Qiskit is not stable for release in an external Python package, because we haven't made any guarantees about the stability of the C interface.

## Requirements

Start by ensuring you have installed the [Qiskit C API](/guides/install-c-api). Next, install the Qiskit Python interface, as follows:
```bash
pip install -r requirements.txt -c constraints.txt
pip install .
```

## Define the C extension

> There are various options to write a C extension for Python. This guide starts with an approach that uses Python's built-in [`ctypes`](https://docs.python.org/3/library/ctypes.html) module, because it requires minimal overhead.  In the next section, the [the Manual C extension](#manual-c-extension) section provides an example of manually building the C extension using Python's C API.

Reviewer: What do you mean by "overhead" here?  The runtime overhead of `ctypes` is higher than a native C extension.

Author: That using `ctypes` requires less boilerplate code, but indeed it's slower so we use the manual extension below. Maybe something like
```suggestion
There are various options to write a C extension for Python. For simplicity, this guide starts with an approach that uses Python's built-in [`ctypes`](https://docs.python.org/3/library/ctypes.html) module.  In the next section, the [the Manual C extension](#manual-c-extension) section provides an example of manually building the C extension using Python's C API for less runtime overhead.
```

correctly across C and Python.
```bash
python setup.py build_rust --inplace --release
```
The shared library is called ``_accelerate.<platform-specific-part>``. Find its location and name as follows:
```bash
QKLIB=$(python -c "import os; import qiskit; print(os.path.dirname(qiskit._accelerate.__file__))")
QKNAME=$(python -c "import os; import qiskit; print(os.path.basename(qiskit._accelerate.__file__))")
```

You will need to know the location of the environments Python includes (``Python.h``) and libraries (``libpython.<suffix>``).
These can, for example, be identified with
```bash
PYINCLUDE=$(python -c "import sysconfig; print(sysconfig.get_path('include'))")
PYLIB=$(python -c "import sysconfig; print(sysconfig.get_config_var('LIBDIR'))")
PYNAME=$(find $PYLIB -maxdepth 1 -name "libpython*" | grep -oE "[^/]+$" | grep -oE "python[0-9]+\.[0-9]+" || echo "python")
```
(If you already know these locations and names, you can also just set them directly.)

### Linux

Linking can differ among platforms. The following describes a solution for linkers
supporting libraries with arbitrary names, using the `-l:` flag. See below if your linker

> requires the library to be called `lib<something>`.

Reviewer: The colon in `-l:<pathspec>` is specific to GNU `ld`.  Using Linux isn't a sufficient or necessary condition for the GNU toolchain to be in use, it's just that Linux systems often distribute the GNU toolchain by default, and other platforms tend not to.

Author: So maybe we can just say to use `-l:` if the linker supports it and otherwise use the symlink solution?

    PyStatus status;
    PyConfig config;
    PyConfig_InitPythonConfig(&config);

    // Add a built-in module, before Py_Initialize.
    if (PyImport_AppendInittab("cextension", PyInit_cextension) == -1) {
        fprintf(stderr, "Error: could not extend in-built modules table\n");
        exit(1);
    }

    // Pass argv[0] to the Python interpreter.
    status = PyConfig_SetBytesString(&config, &config.program_name, argv[0]);
    if (PyStatus_Exception(status)) {
        goto exception;
    }

    // Initialize the Python interpreter.
    status = Py_InitializeFromConfig(&config);
    if (PyStatus_Exception(status)) {
        goto exception;
    }
    PyConfig_Clear(&config);

    // Import the module.
    PyObject *pmodule = PyImport_ImportModule("cextension");
    if (!pmodule) {
        PyErr_Print();
        fprintf(stderr, "Error: could not import module 'cextension'\n");
    }

    return 0;

exception:
    PyConfig_Clear(&config);
    Py_ExitStatusException(status);
}
```
To compile a shared library, link both the Python and Qiskit libraries, as described in
the [Build](#build) section above. The Python script then does not need `ctypes`

> but can directly import `cextension` module:

Reviewer: Well, you'd need to ensure that the `cextension` module is on the Python path somewhere.

print(obs)
```

## Build

First, you have to build the Qiskit Python extension. This includes the C symbols so you can access both interfaces via the same shared library. This is important to ensure data can be passed
correctly across C and Python.
```bash
python setup.py build_rust --inplace --release
```
The shared library is called ``_accelerate.<platform-specific-part>``, whose location and name we can find as
```bash
QKLIB=$(python -c "import os; import qiskit; print(os.path.dirname(qiskit._accelerate.__file__))")
QKNAME=$(python -c "import os; import qiskit; print(os.path.basename(qiskit._accelerate.__file__))")
```

We will need to know the location of the environments Python includes (``Python.h``) and libraries (``libpython.<suffix>``).
These can, for example, be identified via
```bash
PYINCLUDE=$(python -c "import sysconfig; print(sysconfig.get_path('include'))")
PYLIB=$(python -c "import sysconfig; print(sysconfig.get_config_var('LIBDIR'))")
PYNAME=$(find $PYLIB -maxdepth 1 -name "libpython*" | grep -oE "[^/]+$" | grep -oE "python[0-9]+\.[0-9]+" || echo "python")
```
(If you just know these locations and names, you can also just set them directly.)

### Linux

> Linking can slightly differ between platforms, here we first describe a solution for linkers

Reviewer 2: ```suggestion
Linking can differ among platforms. The following describes a solution for linkers
```

supporting libraries with arbitrary names, using the `-l:` flag. See below if your linker
requires the library to be called `lib<something>`.

> We can build the extension specifying the full name of the `_accelerate` library:

Reviewer 2: ```suggestion
You can build the extension specifying the full name of the `_accelerate` library:
```

```bash
gcc extension.c -fpic -shared -o cextension.so \
  -I/path/to/dist/c/include -L$QKLIB -l:$QKNAME \
  -I$PYINCLUDE -L$PYLIB -l$PYNAME
```

> And then simply ``python main.py`` to run the Python program.

Reviewer 2: ```suggestion
Then,  simply enter ``python main.py`` to run the Python program.
```

### MacOS

An alternative to using the exact library name with `-l:` is to symlink the `_accelerate` library

> to the desired name. This solution works on Mac (and also Linux if you replace all occurences of `.dylib` with `.so`).

Reviewer 2: Should the Linux workaround be mentioned in the Linux section above?

Author: We shouldn't need it on linux, the above solution should always work there 🙂 It would just technically be possible to also use this symlink approach, but it's a suboptimal solution that we wouldn't want to encourage, I think 😄

Reviewer 2: got it, thanks!

> To include the `_accelerate` shared library we symlink it to the linkers expected format of `lib<library name>.<suffix>`,

Reviewer 2: ```suggestion
To include the `_accelerate` shared library, symlink it to the linker's expected format of `lib<library name>.<suffix>`:
```

```bash
ln -s $QKLIB/$QKNAME $QKLIB/libqiskit.dylib
```
which allows to use `qiskit` as library name:
```bash
gcc extension.c -fpic -shared -o qiskit_cextension.so \
  -I/path/to/dist/c/include -L$QKLIB -lqiskit \
  -I$PYINCLUDE -L$PYLIB -l$PYNAME
```

> Then simply ``python main.py`` to run the Python program.

Reviewer 2: ```suggestion
Then, simply enter ``python main.py`` to run the Python program.
```

## Manual C extension

> Instead of using `ctypes`, it is possible to manually build an extension for Python using [Python's
> C API](https://docs.python.org/3/c-api/intro.html) directly. This has the potential to be
> faster than using `ctypes`, albeit requires more effort to implement.
> In the following we provide a brief example on how this can be achieved

Reviewer 2: ```suggestion
Instead of using `ctypes`, it is possible to manually build an extension for Python using [Python's
C API](https://docs.python.org/3/c-api/intro.html) directly. This has the potential to be
faster than using `ctypes`, though it requires more effort to implement.
The following code is a brief example on how to achieve this.
```

    // add the term 2 * (X0 Y1 Z2) to the observable
    complex double coeff = 2;  // the coefficient
    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};  // bit terms: X Y Z
    uint32_t indices[3] = {0, 1, 2};  // indices: 0 1 2
    QkObsTerm term = {coeff, 3, bit_terms, indices, num_qubits};
    qk_obs_add_term(obs, &term);  // append the term

    return obs;
}

/// Define the Python function, which will internally build the QkObs using the
/// C function defined above, and then convert the C object to the Python equivalent:
/// a SparseObservable, handled as PyObject.
static PyObject *cextension_build_observable(PyObject *self, PyObject *args) {
    // At this point, ``args`` could be parsed for arguments. See PyArg_ParseTuple for details.
    QkObs *obs = build_observable();  // call the C function to build the observable
    PyObject *py_obs = qk_obs_to_python(obs);  // convert QkObs to the Python-equivalent
    return py_obs;
}

/// Define the module methods.
static PyMethodDef CExtMethods[] = {
    {"build_observable", cextension_build_observable, METH_VARARGS, "Build an observable."},
    {NULL, NULL, 0, NULL}, // sentinel
};

> /// Define the module, which we will call ``cextension``.

Reviewer 2: ```suggestion
/// Define the module, which here is called ``cextension``.
```

    -1,           // keep the module state in global variables
    CExtMethods,
};

PyMODINIT_FUNC PyInit_cextension(void) { return PyModule_Create(&cextension); }

int main(int argc, char *argv[]) {
    PyStatus status;
    PyConfig config;
    PyConfig_InitPythonConfig(&config);

    // Add a built-in module, before Py_Initialize.
    if (PyImport_AppendInittab("cextension", PyInit_cextension) == -1) {
        fprintf(stderr, "Error: could not extend in-built modules table\n");
        exit(1);
    }

    // Pass argv[0] to the Python interpreter.
    status = PyConfig_SetBytesString(&config, &config.program_name, argv[0]);
    if (PyStatus_Exception(status)) {
        goto exception;
    }

    // Initialize the Python interpreter.
    status = Py_InitializeFromConfig(&config);
    if (PyStatus_Exception(status)) {
        goto exception;
    }
    PyConfig_Clear(&config);

    // Import the module.
    PyObject *pmodule = PyImport_ImportModule("cextension");
    if (!pmodule) {
        PyErr_Print();
        fprintf(stderr, "Error: could not import module 'cextension'\n");
    }

    return 0;

exception:
    PyConfig_Clear(&config);
    Py_ExitStatusException(status);
}
```

> To compile a shared library, we link both the Python and Qiskit libraries, as described in
> the [Building](#building) section above. The Python script then does not need `ctypes`

Reviewer 2: ```suggestion
To compile a shared library, link both the Python and Qiskit libraries, as described in
the [Build](#build) section above. The Python script then does not need `ctypes`
```

---

> title: Extending Qiskit in Python with C
> description: How to build a C extension for your Qiskit workflow in Python
> ---
> 
> # A little C with your Python
> 
> To accelerate your Qiskit Python programs with C, we can use Qiskit's C extension for Python.
> This requires additional steps to the standalone C usage, see the [respective guide](./install-c-api).

Reviewer 2: ```suggestion
title: Extend Qiskit in Python with C
description: How to build a C extension for your Qiskit workflow in Python
---

# Extend Qiskit in Python with C

To accelerate your Qiskit Python programs with C, you can use the Qiskit C extension for Python.
This requires additional steps to the standalone C usage; for more details, see the [Install the Qiskit C API guide](/guides/install-c-api).
```

Reviewer 2: I love the whimsy of "A little C with your Python" 🤣 but there are few reasons why to change it - gotta make it easily translatable, match the tone of the rest of the guides, and match the title in the metadata/toc :)

Author: It's a reference to how Rust writes about it: https://docs.rust-embedded.org/book/interoperability/c-with-rust.html 😄 So there's no way we can have a little playful vocabulary in the titles? 😛

Reviewer 2: I would love it, but it would break down in translation, which is of course a GA requirement :(

<Admonition type="note">

>   These instructions have only been tested on unix-based systems. Windows instructions are in progress._

Reviewer 2: ```suggestion
  These instructions have only been tested on UNIX-based systems. Windows instructions are in progress._
```

Author: I should've clarified this: it's only an internal comment -- the final version that we publish should have tested windows instructions too 🙂 but we can also keep this comment as if it were to be published and remove it once we add windows instructions

</Admonition>

## Requirements

> As base, we need the same requirements as the [C standalone usage](./install-c-api).
> In addition we need to install Qiskit's Python interface,

Reviewer 2: ```suggestion
Start by ensuring you have installed the [Qiskit C API](/guides/install-c-api). Next, install the Qiskit Python interface, as follows:
```

>   These instructions have only been tested on UNIX-based systems. Windows instructions are in progress._

Reviewer 3: None of our supported platforms are actually derived or based on UNIX.

```suggestion
  These instructions have only been tested on UNIX-like systems. Windows instructions are in progress._
```

```bash
pip install -r requirements.txt -c constraints.txt
pip install .
```

## Defining the C extension

There are various options to write a C extension for Python. Here, we will describe an approach
using Python's built-in [`ctypes`](https://docs.python.org/3/library/ctypes.html) module
which requires minimal overhead. See [below](#manual-c-extension) for another example which is
manually building the C extension using Python's C API.

As example, assume we write a C function to build an observable and would like to return
it to Python. We can convert a C-side `QkObs*` to a Python-side `SparseObservable` object
(both are backed by the same Rust data structure), using the provided converter
`qk_obs_to_python`:
```c
// file: extension.c
#define PY_SSIZE_T_CLEAN
#include <Python.h>  // include Python header for access to PyObject
#define QISKIT_C_PYTHON_INTERFACE  // enable C->Python conversion functions
#include <qiskit.h>

PyObject *build_observable(void) {
    QkObs *obs = qk_obs_zero(100);
    // build the observable ...
    PyObject *pyobj = qk_obs_to_python(obs); // convert to Qiskit's Python ``SparseObservable``
    qk_obs_free(obs);
    return pyobj;
}
```

> Below we will show how to compile this into a shared library, say `qiskit_cextension.so`.
> Once this is done, we will be able to call the C program from Python as

Reviewer 2: ```suggestion
The following demonstrates how to compile this into a shared library - for example, `qiskit_cextension.so`.
Once this is done, you can call the C program from Python:
```

```python
# file: main.py
import qiskit
import ctypes

> # Load the extension, ensuring the global interpreter lock (GIL) is acquired for function calls,
> # which we need for the C->Python object conversion.
> lib = ctypes.PyDLL("/path/to/qiskit_cextension.so")
> lib.build_observable.argtypes = None  # set argument types to the function
> lib.build_observable.restype = ctypes.py_object  # set return type
> 
> # now we can directly call the function

Reviewer 2: ```suggestion
# Load the extension, ensuring the global interpreter lock (GIL) is acquired for function calls,
# which you need for the C->Python object conversion.
lib = ctypes.PyDLL("/path/to/qiskit_cextension.so")
lib.build_observable.argtypes = None  # set argument types to the function
lib.build_observable.restype = ctypes.py_object  # set return type

# now you can directly call the function
```

Author: everyone needs it 😛  maybe just "which is needed"?

Reviewer 2: IBM Style can be a pain 🤣  the guidance is not to use "we" and instead use the second-person. To say "which is needed" puts it in passive voice, which the guidance also doesn't like :-/  But I think that even in saying "you", it is clear that this means everyone. @beckykd do you have thoughts??

Reviewer 4: I agree that "you" is the preferred term.  Putting it in passive voice is less desirable, but "we" is the least desirable of all :)

obs = lib.build_observable()
print("SparseObservable instance?", isinstance(obs, qiskit.quantum_info.SparseObservable))
print(obs)
```

> ## Building
> 
> First, we have to build Qiskit's Python extension. This will include the C symbols so we can
> access both interfaces via the same shared library. This is important to ensure data ensured
> across C and Python has the same ABI.

Reviewer 2: ```suggestion
## Build

First, you have to build the Qiskit Python extension. This includes the C symbols so you can
access both interfaces via the same shared library. This is important to ensure data ensured
across C and Python has the same ABI.
```

> across C and Python has the same ABI.

Reviewer 2: What is ABI? If a new acronym, we should spell it out first time.

```bash
python setup.py build_rust --inplace --release
```

> The shared library is called ``_accelerate.<platform-specific-part>``, whose location and name we can find as

Reviewer 2: ```suggestion
The shared library is called ``_accelerate.<platform-specific-part>``. Find its location and name as follows:
```

> The shared library is called ``_accelerate.<platform-specific-part>``, whose location and name we can find as
> ```bash
> QKLIB=$(python -c "import os; import qiskit; print(os.path.dirname(qiskit._accelerate.__file__))")
> QKNAME=$(python -c "import os; import qiskit; print(os.path.basename(qiskit._accelerate.__file__))")
> ```
> 
> We will need to know the location of the environments Python includes (``Python.h``) and libraries (``libpython.<suffix>``).

Reviewer 2: ```suggestion
The shared library is called ``_accelerate.<platform-specific-part>``. Find its location and name as follows:
```bash
QKLIB=$(python -c "import os; import qiskit; print(os.path.dirname(qiskit._accelerate.__file__))")
QKNAME=$(python -c "import os; import qiskit; print(os.path.basename(qiskit._accelerate.__file__))")
```

You will need to know the location of the environments Python includes (``Python.h``) and libraries (``libpython.<suffix>``).
```

> We will need to know the location of the environments Python includes (``Python.h``) and libraries (``libpython.<suffix>``).

Reviewer 2: ```suggestion
You will need to know the location of the environments Python includes (``Python.h``) and libraries (``libpython.<suffix>``).
```

> These can, for example, be identified via

Reviewer 2: ```suggestion
These can, for example, be identified with
```

```bash
PYINCLUDE=$(python -c "import sysconfig; print(sysconfig.get_path('include'))")
PYLIB=$(python -c "import sysconfig; print(sysconfig.get_config_var('LIBDIR'))")
PYNAME=$(find $PYLIB -maxdepth 1 -name "libpython*" | grep -oE "[^/]+$" | grep -oE "python[0-9]+\.[0-9]+" || echo "python")
```

> (If you just know these locations and names, you can also just set them directly.)

Reviewer 2: ```suggestion
(If you already know these locations and names, you can also just set them directly.)
```

---
title: Extend Qiskit in Python with C
description: How to build a C extension for your Qiskit workflow in Python
---

# Extend Qiskit in Python with C

To accelerate your Qiskit Python programs with C, you can use the Qiskit C extension for Python.
This requires additional steps to the standalone C usage; for more details, see the [Install the Qiskit C API guide](/guides/install-c-api).

<Admonition type="warning">
  The Qiskit C API is still experimental, and has not yet committed to a fully stable programming or 

>   binary interface.  Extension modules built against Qiskit are only guaranteed to work against 
>   the exact version of Qiskit used in the build, though there are no plans to deliberately break 
>   either interface.

Author: ```suggestion
  binary interface.  Extension modules built against Qiskit are only guaranteed to work against 
  the version of Qiskit used in the build.
```

This requires additional steps to the standalone C usage; for more details, see the [Install the Qiskit C API guide](/guides/install-c-api).

<Admonition type="warning">

>   Qiskit's C API is still experimental, and has not yet committed to a fully stable programming or 
>   binary interface.  Extension modules built against Qiskit are only guaranteed to work against 
>   the exact version of Qiskit used in the build, though we do not plan to deliberately break 

Reviewer 2: ```suggestion
  The Qiskit C API is still experimental, and has not yet committed to a fully stable programming or 
  binary interface.  Extension modules built against Qiskit are only guaranteed to work against 
  the exact version of Qiskit used in the build, though there are no plans to deliberately break 
```
